tests/test_mixins.py

```diff
@@ -0,0 +1,204 @@
+    'coupon-management' feature role in this particular system.
+    """
+
+    @mock.patch('edx_rbac.mixins.utils.contexts_accessible_from_request', return_value=set())
+    @mock.patch('edx_rbac.mixins.utils.contexts_accessible_from_database', return_value=set())
+    # pylint: disable=unused-argument
+    def test_accessible_contexts_superuser(self, mock_contexts_from_request, mock_contexts_from_database):
+        """
+        Superusers get access to all contexts if the `superusers_can_access_anything` switch is on.
+        """
+        viewset = ToyViewSet()
+        viewset.request.user.is_superuser = True
+
+        assert {ALL_ACCESS_CONTEXT} == viewset.accessible_contexts
+
+    @mock.patch('edx_rbac.mixins.utils.contexts_accessible_from_request', return_value=set())
+    @mock.patch('edx_rbac.mixins.utils.contexts_accessible_from_database', return_value=set())
+    # pylint: disable=unused-argument
+    def test_accessible_contexts_superuser_not_granted_all_contexts(
+            self, mock_contexts_from_request, mock_contexts_from_database
+    ):
+        """
+        Toggling the `superusers_can_access_anything` switch off means superusers
+        don't automatically get access to all contexts.
+        """
+        viewset = ToyViewSet()
+        viewset.superusers_can_access_anything = False
+        viewset.request.user.is_superuser = True
+
+        assert set() == viewset.accessible_contexts
+
+    @ddt.data(
+        (set(), set()),
+        ({'context-a'}, {'context-b'}),
+        ({'context-a', 'context-b'}, set()),
+        (set(), {'context-a', 'context-b'}),
+    )
+    @ddt.unpack
+    def test_accessible_contexts_non_superusers(self, contexts_from_jwt, contexts_from_db):
+        """
+        For non-superusers, `contexts_accessible_from_request` and `contexts_accessible_from_database`
+        determine the set of accessible contexts for a user.
+        """
+        viewset = ToyViewSet()
+        viewset.request.user.is_superuser = False
+
+        with mock.patch(
+            'edx_rbac.mixins.utils.contexts_accessible_from_request', autospec=True, return_value=contexts_from_jwt
+        ) as mock_contexts_accessible_from_request, mock.patch(
+            'edx_rbac.mixins.utils.contexts_accessible_from_database', autospec=True, return_value=contexts_from_db
+        ) as mock_contexts_accessible_from_database:
+
+            assert contexts_from_jwt | contexts_from_db == viewset.accessible_contexts
+            assert ALL_ACCESS_CONTEXT not in viewset.accessible_contexts
+
+            mock_contexts_accessible_from_request.assert_called_once_with(
+                viewset.request, viewset.allowed_roles
+            )
+            mock_contexts_accessible_from_database.assert_called_once_with(
+                viewset.request.user, viewset.allowed_roles, viewset.role_assignment_class
+            )
+
+    def test_check_permissions_list_superuser(self):
```

Comment:
It seems like `test_check_permissions_list_superuser`, `test_check_permissions_list_staff_never_forbidden`, and `test_check_permissions_list_staff_can_be_forbidden` all pretty much share the same logic except for changing some flags around.

Could these tests be collapsed into one with the help of `@ddt`?

Or, if there's value in having the tests broken out to have a describe test name associated with them, feel free to ignore :)

Reply:
I think the first two can be collapsed.  We could add the third in, too, but it would be a little bit gross because of the exception assertion.